OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -19,47 +19,165 @@ namespace OpenRA.Mods.Common.Widgets
 {
 	public class SupportPowerTimerWidget : Widget
 	{
+		public readonly int XObserver = 10;
+		public readonly int YObserver = 10;
+		public enum TimerOrder { AlliesFirst, EnemiesFirst }
+		public readonly TimerOrder Order = TimerOrder.EnemiesFirst;
+		[Translate] public readonly string HeaderAllies = "Allies";
+		[Translate] public readonly string HeaderEnemies = "Enemies";
 		public readonly string Font = "Bold";
 		public readonly string Format = "{0}: {1}";
-		public readonly TimerOrder Order = TimerOrder.Descending;
 
-		readonly IEnumerable<SupportPowerInstance> powers;
-		Pair<string, Color>[] texts;
+		readonly IEnumerable<SupportPowerInstance> allPowers;
+		Player[] allPlayers, playersSelf, playersAllies, playersEnemies;
+		Pair<string, Color>[] allTexts, textsSelf, textsAllies, textsEnemies;
+		delegate void Del();
+		Del callDrawTimers = () => { };
+		readonly World world;
+		int lastTick;
 
 		[ObjectCreator.UseCtor]
 		public SupportPowerTimerWidget(World world)
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
+			allPowers = world.ActorsWithTrait<SupportPowerManager>()
 				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
 				.SelectMany(s => s.Trait.Powers.Values)
 				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+
+			this.world = world;
+		}
+
+		Pair<string, Color> GetText(SupportPowerInstance spi)
+		{
+			var time = WidgetUtils.FormatTime(spi.RemainingTime, false);
+			var text = Format.F(spi.Info.Description, time);
+			var color = !spi.Ready || Game.LocalTick % 50 < 25 ? spi.Instances[0].Self.Owner.Color.RGB : Color.White;
+			return Pair.New(text, color);
 		}
 
 		public override void Tick()
 		{
```

Comment:
There's still a lot of duplication and unclear logic flow in this method, which will need to be cleaned up before we can merge this.